### PR TITLE
fix(cloud-ai-sdk): cancel telemetry streams with chat runs

### PR DESCRIPTION
**Pull request by Kinfe123**
## Problem

Cancelling a Cloud chat before its first token did not cancel the response stream for custom transports that rely on stream cancellation. The transport stream was split for telemetry, but cancellation only reached the branch returned to the chat runtime, leaving the telemetry branch waiting for a token.

## Root cause

`observeTelemetryStream()` returned one side of a `ReadableStream.tee()` while consuming the other side in the background. A tee source is cancelled only after both branches are cancelled, so cancelling the chat branch alone never propagated to the source.

## Change

Forward chunks through one owned source reader. It reads eagerly only until the first token to preserve arrival-time telemetry, then follows downstream backpressure. Cancellation now reaches the source with the caller's original reason, and every completion path releases the reader lock.

## Verification

- Added a bounded pre-token cancellation regression that hangs before this change and confirms the original reason reaches the source after it.
- Added post-token cancellation and source-error forwarding coverage.
- Extended the telemetry completion test to pin eager first-token timing and source lock release.
- `pnpm -C packages/cloud-ai-sdk test`: 134 tests passed.
- `aui-build` for `@assistant-ui/cloud-ai-sdk` passed.
- Focused oxlint and changeset validation passed.

## Public surface

`@assistant-ui/cloud-ai-sdk` runtime behavior only. No exports or API signatures changed. Includes a patch changeset.

### CODE REVIEW

File: packages/cloud-ai-sdk/src/core/CloudChatCore.ts [hunk 1]
```diff
@@ -319,10 +319,21 @@ export class CloudChatCore {
     timing: ActiveTelemetryTiming,
   ): ReadableStream<UIMessageChunk> {
     const [chatStream, timingStream] = stream.tee();
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/cloud-ai-sdk/src/core/CloudChatCore.ts]: Addressed in a237800be. I replaced the tee and manual reader wrapper with the identity TransformStream, and updated the timing test to consume the transformed stream.

File: packages/cloud-ai-sdk/src/core/CloudChatCore.ts [hunk 2]
```diff
@@ -332,8 +343,46 @@ export class CloudChatCore {
...
+            controller.enqueue(value);
+          }
+        } catch (error) {
+          releaseChatReader();
+          controller.error(error);
+        }
+      },
+      async cancel(reason) {
+        await Promise.all([stopChat(reason), stopTiming(reason)]);
+      },
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/cloud-ai-sdk/src/core/CloudChatCore.ts:386]: Addressed in a237800be. Native transform cancellation now forwards the caller reason unchanged, and the regression test pins the exact reason received by the source.

File: packages/cloud-ai-sdk/src/core/CloudChatCore.test.ts [hunk 3]
```diff
@@ -218,6 +218,25 @@ describe("CloudChatCore", () => {
...
+    vi.spyOn(core, "persist").mockResolvedValue(undefined);
+
+    const stream = await core
+      .createTransport("chat-1", registry)
+      .sendMessages({ messages: [] } as never);
+    const reason = new Error("stopped");
+    await stream.cancel(reason);
+
+    expect(cancel).toHaveBeenCalledOnce();
+  });
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/cloud-ai-sdk/src/core/CloudChatCore.test.ts:273]: The custom wrapper and its post-token/reader-release paths are gone in a237800be. The existing telemetry test now consumes the identity transform and still pins firstTokenMs, while the focused cancellation test covers the original pre-token failure and exact cancellation reason.

File: packages/cloud-ai-sdk/src/core/CloudChatCore.test.ts [hunk 4]
```diff
@@ -218,6 +218,26 @@ describe("CloudChatCore", () => {
     );
   });
 
+  it("cancels the transport stream while waiting for the first token", async () => {
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/cloud-ai-sdk/src/core/CloudChatCore.test.ts:223]: Addressed in 76580255b. The existing timing test now proves eager first-token observation and normal completion, including source reader lock release. A separate regression covers cancellation after the first token.

File: packages/cloud-ai-sdk/src/core/CloudChatCore.test.ts [hunk 5]
```diff
@@ -218,6 +218,26 @@ describe("CloudChatCore", () => {
...
+      baseTransport: { sendMessages, reconnectToStream: vi.fn() },
+    });
+    vi.spyOn(core, "ensureThreadId").mockResolvedValue("thread-1");
+    vi.spyOn(core, "persist").mockResolvedValue(undefined);
+
+    const stream = await core
+      .createTransport("chat-1", registry)
+      .sendMessages({ messages: [] } as never);
+    const reason = new Error("stopped");
+    await stream.cancel(reason);
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/cloud-ai-sdk/src/core/CloudChatCore.test.ts:269]: Addressed in 76580255b. The test now waits only up to 250 ms for the source cancel callback before awaiting cancellation, so the old tee regression fails quickly instead of reaching Vitest’s default timeout.

File: packages/cloud-ai-sdk/src/core/CloudChatCore.ts [hunk 6]
```diff
@@ -318,22 +318,19 @@ export class CloudChatCore {
...
-          timing.firstTokenMs = Date.now() - timing.startedAt;
-          return;
-        }
-      }
-    };
-    void readUntilFirstToken()
-      .catch(() => {})
-      .finally(() => reader.cancel().catch(() => {}));
-    return chatStream;
+    return stream.pipeThrough(
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/cloud-ai-sdk/src/core/CloudChatCore.ts]: Addressed in 76580255b. The implementation now eagerly forwards through a single source reader until the first token, then resumes normal pull-based backpressure. Cancellation reaches that reader with the original reason, and the timing test waits before consuming to pin arrival-time observation.

File: packages/cloud-ai-sdk/src/core/CloudChatCore.ts [hunk 7]
```diff
@@ -318,22 +318,70 @@ export class CloudChatCore {
...
+        eagerRead = readUntilFirstToken(controller);
+      },
+      async pull(controller) {
+        if (eagerRead) {
+          await eagerRead;
+          eagerRead = undefined;
+        } else {
+          await forwardNext(controller);
+        }
+      },
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/cloud-ai-sdk/src/core/CloudChatCore.ts:376]: Pinned in 7b2a59cbc: the token is enqueued at clock 125, the test waits for eager observation, then advances to 150 before the consumer reads while still expecting firstTokenMs 25. The identity transform records 50 in that delayed-consumer case, so the test now captures the concrete telemetry difference that justifies the eager reader.

File: packages/cloud-ai-sdk/src/core/CloudChatCore.test.ts [hunk 8]
```diff
@@ -194,8 +194,9 @@ describe("CloudChatCore", () => {
     controller!.enqueue({ type: "text-start", id: "part-1" });
     controller!.enqueue({ type: "text-delta", id: "part-1", delta: "hi" });
     controller!.close();
-    await stream.getReader().read();
     await new Promise((resolve) => setTimeout(resolve, 0));
+    await stream.pipeTo(new WritableStream());
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/cloud-ai-sdk/src/core/CloudChatCore.test.ts:199]: Addressed across 7b2a59cbc and 76e2afce7. The timing test now advances the clock before downstream consumption, and the new source-error regression asserts the exact error reaches the consumer and source.locked is false afterward.

File: packages/cloud-ai-sdk/src/core/CloudChatCore.ts [hunk 9]
```diff
@@ -318,22 +318,70 @@ export class CloudChatCore {
     stream: ReadableStream<UIMessageChunk>,
     timing: ActiveTelemetryTiming,
   ): ReadableStream<UIMessageChunk> {
-    const [chatStream, timingStream] = stream.tee();
-    const reader = timingStream.getReader();
-    const readUntilFirstToken = async () => {
-      while (true) {
+    const reader = stream.getReader();
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/cloud-ai-sdk/src/core/CloudChatCore.ts:322]: Added the invariant comment in 93c945c5b so the eager first-token read remains tied to arrival-time telemetry.

File: packages/cloud-ai-sdk/src/core/CloudChatCore.ts [hunk 10]
```diff
@@ -318,22 +318,70 @@ export class CloudChatCore {
...
+          await eagerRead;
+          eagerRead = undefined;
+        } else {
+          await forwardNext(controller);
+        }
+      },
+      async cancel(reason) {
+        cancelled = true;
+        try {
+          await reader.cancel(reason);
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/cloud-ai-sdk/src/core/CloudChatCore.ts]: Valid edge case. 93c945c5b now skips source cancellation after the reader has already been released, and adds a regression test for cancelling a wrapper with queued non-token chunks after source EOF.
